### PR TITLE
Spark: Exclude netty from spark runtime modules

### PR DESCRIPTION
**Pull request by pan3793**
The unexpected `io.netty.util` comes from `netty-common` exists in spark runtime jars, we should relocate it to avoid classes conflict.

image

### CODE REVIEW

**Comment by rdblue** [comment]
@pan3793, if there is something shared between Iceberg and Spark, we generally prefer to exclude it from the runtime Jar. Is it possible to make sure the version matches and remove it from the runtime?

**Comment by pan3793** [comment]
@rdblue The netty deps come from arrow, Spark and Iceberg doesn't guarantee to use the same version of arrow, but for netty, I think it's good to share it, because both arrow and spark use netty `4.1.x.Final`, netty stays `4.1.x.Final` for long time, and I never hit class/jar conflict cases from netty.

**Comment by pan3793** [comment]
@rdblue would you please take a look again?

File: spark/v2.4/build.gradle [hunk 1]
```diff
@@ -162,7 +163,6 @@ project(':iceberg-spark:iceberg-spark-runtime') {
     relocate 'org.apache.orc', 'org.apache.iceberg.shaded.org.apache.orc'
     relocate 'io.airlift', 'org.apache.iceberg.shaded.io.airlift'
     // relocate Arrow and related deps to shade Iceberg specific version
-    relocate 'io.netty.buffer', 'org.apache.iceberg.shaded.io.netty.buffer'
```
Comment (rdblue) [inline spark/v2.4/build.gradle:165]: The Spark dependency on netty hasn't changed since Spark 2.4?

Comment (rdblue) [reply spark/v2.4/build.gradle:165]: Looks like there was a 3.9.9.Final dependency in 2.4.0.

Comment (pan3793) [reply spark/v2.4/build.gradle:165]: 2.4.0 bundles `io.netty:netty-all:4.1.17.Final`, see https://mvnrepository.com/artifact/org.apache.spark/spark-core_2.11/2.4.0

Comment (pan3793) [reply spark/v2.4/build.gradle:165]: netty3 and netty4 classes use different package name, they can co-exist.

**Comment by pan3793** [comment]
Kindly ping @rdblue

**Comment by rdblue** [comment]
Thanks, @pan3793!

**Comment by ajantha-bhat** [comment]
@pan3793 , @rdblue : 
I was using latest build from master branch with `iceberg-spark-runtime-3.0_2.12-0.14.0-SNAPSHOT.jar`.
Basic select query failed with the below error.

`Caused by: java.lang.NoSuchMethodError: io.netty.buffer.PooledByteBufAllocatorL.allocate(J)Lio/netty/buffer/UnsafeDirectLittleEndian;`

Screenshot 2022-03-16 at 6 24 46 PM

**Once I revert this Change. Query was success for me.**

**Comment by ajantha-bhat** [comment]
Also tried on our [snapshot jars published](https://repository.apache.org/content/repositories/snapshots/org/apache/iceberg/iceberg-spark-runtime-3.0_2.12/0.14.0-SNAPSHOT/). Same issue.

File: spark/v2.4/build.gradle [hunk 2]
```diff
@@ -121,6 +121,7 @@ project(':iceberg-spark:iceberg-spark-runtime') {
       exclude group: 'org.xerial.snappy'
       exclude group: 'javax.xml.bind'
       exclude group: 'javax.annotation'
+      exclude group: 'io.netty'
```
Comment (rzhang10) [inline spark/v2.4/build.gradle:124]: I feel we should still remove the unnecessary netty dependencies that spark already has on its runtime, and the correct thing to do here is:
```
exclude group: 'io.netty', module: 'netty-common'
```

@rdblue @pan3793 @ajantha-bhat what do you think?

**Pull request merged**